### PR TITLE
Rework hardware acceleration decoder selection 

### PR DESCRIPTION
**Pull request by lizardfish0**
Bit to explain here... 

tl;dr, I needed a way to force the use of a specific `hwaccel`/`decoder` pair. 

This PR addresses two problems:

1. Improve decoder selection logic to only use decoders that are included in `hwaccel_decoders`.

Currently, the script will attempt to use a decoder that matches `<input codec>_<hwaccel name>`, provided it exists in `ffmpeg -decoders`. The first and largest issue with this approach is that this codec might not actually be supported by the underlying hardware. For example, if the input codec is `av1` and you've provided `cuda` as a hwaccel, then you'll need Nvidia 30-series or later to run the `av1_cuvid` decoder.

Additionally, the script will currently append the first valid (exists in ffmpeg) `hwaccel` it finds.  In my case, even though my GPU didn't support `av1`, I'd like to use my CPU's iGPU to perform the decoding. This PR makes that possible.

2. Users might want to select a specific decoder that doesn't fit the `<input codec>_<hwaccel name>` mold.

I wasn't aware of this, but when trying to solve my use-case I learned that there are internal ffmpeg codecs that support hardware acceleration, and there are external ffmpeg codecs specifically built for a single hardware platform. ffmpeg will implicitly use the internal codec unless you tell it otherwise.

 i.e. you can run the implicit `hevc` decoder with `-hwaccel cuda`, or run `hevc_cuvid` with `-hwaccel cuda`. I don't know too much about how these are maintained separately, but I read that sometimes there are differences in implementation that make one more efficient, so perhaps this will be useful to some.

---

Core problem: there is no good way to query ffmpeg to know if a given decoder is actually going to work with the provided hardware. The only one who knows whether it's going to work is the user. 

Might be helpful to run through my situation to understand why this might be useful.

Hardware:
- GPU: 1650 Super w/ Turing NVENC/NVDEC
- CPU: i12400

Previously, I had
```
[Converter]
...


hwaccel-decoders = h264_cuvid, hevc_cuvid, mjpeg_cuvid, mpeg1_cuvid, mpeg2_cuvid, mpeg4_cuvid, vp8_cuvid, vp9_cuvid
hwdevices = 
hwaccel-output-format = cuda:cuvid
...
[Video]
codec = hevc_nvenc, hevc, h265, x265, h264, x264
...
```

This worked well until I ran into an AV1 file, which failed transcoding 
```
[av1 @ 0x55e6c68fdd80] Hardware is lacking required capabilities
[av1 @ 0x55e6c68fdd80] Failed setup for format cuda: hwaccel initialisation returned error.
[av1 @ 0x55e6c68fdd80] Your platform doesn't support hardware accelerated AV1 decoding.
[av1 @ 0x55e6c68fdd80] Failed to get pixel format.
```

My iGPU supports AV1 decoding, so I figured I could wrangle the settings into performing the decode with that. However, the following fails because the script attempts to use `cuda`.
```
hwaccels = cuda, vaapi
```

However, the script searches for `<input codec>_<hwaccel name>` and there is no `av1_vaapi`. The solution would be to use `-hwaccel vaapi` with `-vcodec av1`.
> `vcodec` would actually be optional due to ffmpeg using it implicitly

Thus the second change. 

If you have a cleaner solution I'm happy to work it out, but I think this works pretty well. I explored modifying the way `hwaccel_decoders` works, where we could detect a listed decoder that wasn't valid according to ffmpeg but corresponded to an internal decoder with hardware acceleration. Adding a new setting to manually specify `hwaccel/decoder` pairings seemed much cleaner and less confusing to users.

### CODE REVIEW

File: converter/ffmpeg.py [hunk 1]
```diff
@@ -540,15 +540,25 @@ def hwaccel_decoder(self, video_codec, hwaccel):
...
+                return formats[1].strip().split(" ")
+        else:
+            return []
 
     def decoder_formats(self, decoder):
         prefix = "Supported pixel formats:"
-        formatline = next((line.strip() for line in self._get_stdout([self.ffmpeg_path, '-hide_banner', '-h', 'decoder=%s' % decoder]).split('\n')[1:] if line and line.strip().startswith(prefix)), "")
-        formats = formatline.split(":")
-        return formats[1].strip().split(" ") if formats and len(formats) > 0 else []
+        format_line = next((line.strip() for line in self._get_stdout([self.ffmpeg_path, '-hide_banner', '-h', f"decoder={decoder}"]).split('\n')[1:] if line and line.strip().startswith(prefix)), "")
```
Comment (lizardfish0) [inline converter/ffmpeg.py:554]: These changes avoid an index out-of-bounds-errors when trying to get formats for invalid codecs.

Comment (lizardfish0) [inline converter/ffmpeg.py:554]: These changes avoid an index out-of-bounds-errors when trying to get formats for invalid codecs.

File: resources/mediaprocessor.py [hunk 2]
```diff
@@ -1490,64 +1494,85 @@ def checkDisposition(self, allowed, source):
                 return False
         return True
 
-    # Hardware acceleration options now with bit depth safety checks
-    def setAcceleration(self, video_codec, pix_fmt, codecs=[], pix_fmts=[]):
+    def set_decoder(self, video_codec: str, pix_fmt: str):
```
Comment (lizardfish0) [inline resources/mediaprocessor.py:1497]: Renamed this as its specific to decoders.

File: resources/mediaprocessor.py [hunk 3]
```diff
@@ -1490,64 +1494,85 @@ def checkDisposition(self, allowed, source):
...
+            if device:
+                self.log.debug("Setting hwaccel device to %s." % device)
+                opts.extend(['-init_hw_device', '%s=sma:%s' % (_hwaccel, device)])
+                opts.extend(['-hwaccel_device', 'sma'])
+
+            # Set vcodec
+            opts.extend(['-vcodec', _decoder])
+
+        # If there's a manually specified hwaccel/decoder pairing for this codec, use it.
+        if video_codec in self.settings.hwaccel_decoder_override:
```
Comment (lizardfish0) [inline resources/mediaprocessor.py:1533]: For a specific input codec, users will be able to specify a `hwaccel` and a `decoder` using the format `<codec>:<hwaccel>.<decoder>`.

ex:
```
hwaccel_decoder_override = av1:vaapi.av1
```
Happy to add a bit to the wiki about this.

File: resources/mediaprocessor.py [hunk 4]
```diff
@@ -1490,64 +1494,85 @@ def checkDisposition(self, allowed, source):
...
+                        video_codec,
+                        self.settings.hwoutputfmt.get(hwaccel, hwaccel)
+                    )
+
+                    self.log.debug(f"Target decoder: {target_decoder}")
+                    self.log.debug(f"Target decoder pixel formats: {self.converter.ffmpeg.decoder_formats(target_decoder)}.")
+
+                    is_supported_decoder = target_decoder in codecs[video_codec]['decoders']
+
+                    if is_supported_decoder and target_decoder in self.settings.hwaccel_decoders:
```
Comment (lizardfish0) [inline resources/mediaprocessor.py:1566]: This slightly modifies existing behavior. Only specifying `hwaccels=` in settings will now do nothing. Most examples I've seen of users attempting hardware acceleration on this repo have been specifying their decoders anyways.

File: converter/ffmpeg.py (hunk 1 above)
Comment (mdhiggins) [reply converter/ffmpeg.py:554]: Looks like this was just a mistake on my part, the line 

```py
return formats[1].strip().split(" ") if formats and len(formats) > 0 else []
```

should have read

```py
return formats[1].strip().split(" ") if formats and len(formats) > 1 else []
```

Fixed that with e62addfc7e68009e7dadde2aae2a6e0d9b5e40d4

**Comment by mdhiggins** [comment]
Reviewing this now

Question though, does using `-hwaccel vaapi -vcodec av1` actually use your iGPU?

If your ffmpeg build doesn't have a vaapi decoder for av1 isn't it just falling back to software based on those options? Perhaps my understanding isn't correct but that was my assumption

Also, could you share the ffmpeg command generated for this output

```
[av1 @ 0x55e6c68fdd80] Hardware is lacking required capabilities
[av1 @ 0x55e6c68fdd80] Failed setup for format cuda: hwaccel initialisation returned error.
[av1 @ 0x55e6c68fdd80] Your platform doesn't support hardware accelerated AV1 decoding.
[av1 @ 0x55e6c68fdd80] Failed to get pixel format.
```

And additionally could you share the command generated with your fork?

Seems like I probably need to allow -hwaccel to be added if the decoder list is empty since some people still use that basic config option to get some generic hwaccel

**Comment by lizardfish0** [comment]
> Question though, does using -hwaccel vaapi -vcodec av1 actually use your iGPU?

It does, confirmed with `intel-gpu-top`. I assume it would fall back on software but I believe that's the best option.

Was about to grab you some ffmpeg outputs but it looks like something just changed with the `sonarr-sma` alpine container, and `SMA_USE_REPO` now installs a build of ffmpeg that doesn't have any hardware encoding support.

**Comment by mdhiggins** [comment]
Had a bug with the startup script on sonarr-sma that I just fixed like 2 minutes ago, might want to just do a fresh pull and try again

**Comment by mdhiggins** [comment]
Yeah upon reviewing the decoders, it looks like vaapi is the only one that doesn't respect the codec_hwaccel naming convention for the decoders (though it does for the encoders)
Could probably hard-code that exception

**Comment by lizardfish0** [comment]
meh, I'm now an hour down a rabbit hole discovering why I have a half-finished attempt to put `ffmpeg` behind a thin http api so that we can abstract it from alpine-based images (maybe I'll revisit). My `sonarr-sma` image was in a weird state. I have no idea how I had a build of `ffmpeg` that supported qsv/vaapi/cuda/nvenc (I run a `pull` & `up` pretty regularly, maybe there was some weirdness in my docker stack left over from using the #build tag).

Regardless, off the top of my head I believe the script without any changes generated:
```
ffmpeg -hwaccel cuda -i av1_input.mkv -vcodec hevc_nvenc hevc_output.mkv
```

And now with the changes it generates
```
ffmpeg -hwaccel vaapi -vcodec av1 -i av1_input.mkv -vcodec hevc_nvenc hevc_output.mkv
```

Currently, the script will always set `-hwaccel` to the first entry in `hwaccel` that exists in ffmpeg. Definitely useful to be able to filter that selection by also requiring the decoder to exist in `hwaccel_decoders`. Hard-coding an exception to `vaapi` would definitely solve that issue, but I think there is still value in letting users force a vcodec (if they really wanted to use `-hwaccel cuda -vcodec hevc` instead of `-hwaccel cuda -vcodec hevc_cuvid`). Likely very rare that anyone would use it, but it allows for a lot of control/edge cases.

**Comment by lizardfish0** [comment]
Thoughts on adding support for https://github.com/jellyfin/jellyfin-ffmpeg? Looks like a great source of pre-compiled ffmpeg binaries w/ hwaccel support.

**Comment by henryjb99** [comment]
I know this is a really old pull request and probably out of date, but I wanted to chat about this idea. I am running into this exact problem where AV1 files are attempting to be decoded by ffmpeg using my NVidia 1070TI. However, the decoding fails as the 1070Ti does not support av1_cuvid, but ffmpeg thinks it does.

**Comment by mdhiggins** [comment]
As long as that decoder isn't on your list it shouldn't be applied

As far as the other changes to this, the hardware acceleration implementation of SMA could use a big overhaul I just haven't had time to redo it yet, its on my todo list but intend to incorporate some of the changes from this pull request

**Comment by henryjb99** [comment]
That makes sense. I decided to upgrade my video card to one that does support decoding AV1. That should help me bypass my issue.

I also wanted to say a very heartfelt thank you for this product and all the work you do on it and your other projects. You are helping a lot of people and we appreciate it.
